Support writers without static lifetimes

tofay opened the pull request:
I've been trying to use this crate with https://github.com/containers/ocidir-rs to create container images. Getting that working required two changes to gzp:
* the ability to return the underlying writer from Zwriter::finish
* the ability to use borrowed writers in ParCompressBuilder::from_borrowed_writer

The latter is accomplished by allowing users to provide a `https://doc.rust-lang.org/stable/std/thread/struct.Scope.html` alongside the writer - the writer thread is then spawned using this scope.

Commit:
Zwriter::finish returns the underlying writer

Commit:
add support for borrowed writers

ParCompressBuilder::from_borrowed_writer takes a writer and
a std::thread::Scope reference. The writer thread is spawned using
the scope, so the writer needn't have a 'static lifetime.

sstadick commented:
Hi @tofay - I'm sorry for not seeing this earlier. Do you have an example branch of your fork of this in use that I could look at? I think I get what you're saying, I'm just trying to understand the motivation for wanting this. I'm a bit reticent to add a lifetime parameter to an already fairly complex type, but I'm not against it if the use case is compelling!

Some1and2-XC commented:
> Hi @tofay - I'm sorry for not seeing this earlier. Do you have an example branch of your fork of this in use that I could look at? I think I get what you're saying, I'm just trying to understand the motivation for wanting this. I'm a bit reticent to add a lifetime parameter to an already fairly complex type, but I'm not against it if the use case is compelling!

I think this is a solid change, though I understand the hesitancy of adding lifetimes..

I find the problem with requiring a static lifetime here is that without it, using the output of a compressed stream is somewhat difficult. Constructing a reference to write to that is static is hard when you want to use the resource later. This seems to be also shown in the test included in the first [pull request](https://github.com/sstadick/gzp/pull/55/commits/fca574f6c2f9006b40c8ea772de1f28bbf3e580e) (I can't think of a way to get the same behavior in the current implementation without reconstructing the reference to the file).

I also ran into the problem of getting the data out of the compressor [here](https://github.com/sstadick/gzp/issues/57#issuecomment-2614159063) and the best I could come up with was using a channel to pass the data out which I think comes with some overhead (not benchmarked).

tofay commented:
https://github.com/tofay/gnoci/commit/5ab6d2024c9ab19617d6a91dc1be668612ace9d7 is an example. I'm creating container images with https://github.com/containers/ocidir-rs. That uses cap-std, and the writes of container images are done using a non-static lifetime. To speed up compression I can use gzp, but I can't use the existing `ParCompressBuilder::from_writer` as is. With my branch of gzp I can do all the writing within a thread scope.

sstadick reviewed (commented):
Could you update the CHANGELOG with an entry for this addition?

Thank you for pinging me with the use case example and for persisting in asking about this! The example really helped me understand why it's useful. Let's get this in.

Once this is merged in, I'll take a pass at updating dependencies / generally doing some project maintenance and then make a release.

tofay committed:
Add examples and docs for from_borrowed_writer

tofay commented:
> Could you update the CHANGELOG with an entry for this addition?
> 
> Thank you for pinging me with the use case example and for persisting in asking about this! The example really helped me understand why it's useful. Let's get this in.
> 
> Once this is merged in, I'll take a pass at updating dependencies / generally doing some project maintenance and then make a release.

Thanks! I've now added the changelog entry.

sstadick commented:
Thanks @tofay! I will try to take a look at the failing lints and tests this week. The lints are for sure not something on your end, and I'm pretty sure the failing tests are also not related to your changes.

sstadick commented:
@tofay see https://github.com/tofay/gzp/pull/1

sstadick committed:
fix: resolve failing tests for PR 55 (#1)

sstadick committed:
fix: remove override from CI to respect rust-toolchain.toml

The CI was using 'override: true' which caused it to ignore the
rust-toolchain.toml file and use the latest stable Rust. This resulted
in clippy lints failing due to the new io_other_error lint in newer
Rust versions.

Removing override allows the CI to respect rust-toolchain.toml and use
Rust 1.80.0, making the lints pass.

sstadick committed:
fix: add rustfmt and clippy components to rust-toolchain.toml

The CI needs rustfmt and clippy components to be installed for the
toolchain specified in rust-toolchain.toml.

sstadick commented:
@tofay see v2.0.0

sstadick commented:
@Some1and2-XC see v2.0.0
